Add support for normalization of point sizes

astrofrog opened the pull request:
This PR adds a few attributes that can be used to control whether the point sizes are interpreted as-is from the ``sizeColumn`` or whether they are normalized beforehand using:

```
new_size = (size - size_min) / (size_max - size_min)
```

In many cases, astronomers don't have tables with point sizes already in the table - instead they will have some value, e.g. mass or temperature and will want to map these to actual point sizes. Rather than have to re-generate the table many times, the attributes here allow normalization to be turned on in the SDK, and to control the min/max.

This will allow us to then implement a much more efficient approach for https://github.com/WorldWideTelescope/pywwt/issues/137 in PyWWT - at the moment, we constantly re-generate data tables and send them from Python to JS - with this, we'll be able to just modify the min/max.

Example usage in the JS SDK:

```js
csv = 'a,b,c\r\n266,-29,1\r\n267,-29,3\r\n267,-30,5\r\n266,-30,10\r\n';
layer = wwtlib.LayerManager.createSpreadsheetLayer('Sky', 'My Layer', csv);
layer.set_lngColumn(0);
layer.set_raUnits(1);
layer.set_latColumn(1);
layer.set_sizeColumn(2);
layer.set_scaleFactor(10);
layer.set_pointScaleType(0);
layer.set_normalizeSize(true);
layer.set_normalizeSizeMin(2);
layer.set_normalizeSizeMax(10);
layer.set_normalizeSizeClip(true);
wwt.gotoRaDecZoom(266.5, -29.5, 10, true);
```

The default is to not do any normalization, consistent with the current default behavior. In addition, this doesn't interact in any way with the XML files, which on the plus side means no breakage of tour file compatibility, and on the minus side means that these settings wouldn't be saved in a tour. If we are ok with adding new attributes to tour files, then I can do this.

Note that technically speaking, since the min/max default to 0 and 1, and clip defaults to ``false``, the ``normalizeSize`` option isn't really needed, but having it set to ``false`` does avoid even doing the normalization calculation at all, which might be better for performance (but it probably won't make much difference).

Also I've only implemented this for ``pointScaleType`` set to ``Linear`` or ``Power`` since those cases are straightfoward. ``Log`` normalization is a little trickier since it involves an additional constant (see the equation in [LogStretch](http://docs.astropy.org/en/stable/api/astropy.visualization.LogStretch.html#astropy.visualization.LogStretch) for astropy). And I think ``StellarMagnitude`` shouldn't be scaled/normalized since it has absolute meaning.

pkgw commented on HTML5SDK/wwtlib/Layers/SpreadSheetLayer.cs:
```diff
@@ -2031,6 +2030,108 @@ public int SizeColumn
+        //
+        // The NormalizeSizeClip attribute can be used to determine whether the sizes should
+        // be clipped to the range [0:1]. At this time, normalization is only applied if
+        // PointScaleTypes is Linear or Power.
```
I'm wondering if we can figure out something better for the clipping behavior.

First of all, regardless of whether the point size is measured in pixels or degrees, it's not obvious to me that a value of 1 is a natural maximum to clip to. If the final computed size can be interpreted as a number of pixels (caveat: I do not know if this is indeed the case), clipping to a range of 0-1 seems quite restrictive.

The other thing is that clipping behavior at 0 and 1 seems pretty asymmetrical. My guess would be that negative radii would render as their absolute values, i.e. a radius of -1 is the same as a radius of 1. This could be hard to spot as a user. Meanwhile, radii above 1 almost certainly have continuous behavior.

I think I'd suggest that rather than a boolean clip flag, there be a floating-point `ClipMax` attribute above which values are always clipped. It could default to +Inf. Radii below 0 would always be clipped to 0.

In such a model, it might make sense to rename the `NormalizeSizeMax` parameter to `NormalizeSizeUnity`, or something like that.

astrofrog replied:
The way I was using this currently was to use the min/max to get values in the range [0:1], then using the clip on that, and then use the ``scaleFactor`` setting to scale the points to whatever absolute size I want, so e.g. using ``scaleFactor`` of 10 would lead to clipped values in the range [0:10]. This does mean that the clip value only makes sense when the min/max are also specified, and requires ``scaleFactor`` to be specified.

pkgw replied:
Oh, the `scaleFactor` also gets applied in this chain? Then yes, this makes more sense.

astrofrog replied:
Yes it gets applied, not in the methods I wrote, but in the next step.

pkgw commented on HTML5SDK/wwtlib/Layers/SpreadSheetLayer.cs:
```diff
@@ -2031,6 +2030,108 @@ public int SizeColumn
+            else
+            {
+                new_value = value;
+            }
```
Style bikeshedding: I think it's good to "resolve" `if` statements briefly, if possible, so that the reader doesn't have to remember what condition is being considered. In this particular case, what I'm getting at is:

```
if(!NormalizeSize)
    return value;

float new_value = (value - NormalizeSizeMin)/ ....;
if (!NormalizeSizeClip)
    return new value;

if (new_value < 0)
...
return new_value;
```

This reduces rightward drift and, I'd like to think, cognitive load on the reader, since they can "retire" different conditions in the terse `if` clauses.

astrofrog replied:
Should be fixed in https://github.com/WorldWideTelescope/wwt-web-client/pull/237/commits/7d7a657fa2474fe4fed69205b8b6894d13311268 - let me know if that's not what you had in mind.

pkgw commented:
As for XML serialization ... if new attributes are added, older clients won't know about them, and will just read the values in the data table as-is, and give results different than a client that knows about those attributes

I'm torn because I want to get these things right, but if we take a strict approach then the existing XML file format becomes an extremely restrictive straitjacket.

What if the XML export wrote out a synthetic size column that applied the linear transform and/or clipping? Is there any chance that that would be relatively straightforward to implement? I'm guessing not ...

astrofrog commented:
@astrojonathan @pkgw - I've now adjusted this PR so that the layer settings are saved in a backward-compatible fashion in a way that can be read in by an older WWT web client version or the desktop WWT version. Here's a preview of creating the layer in the web client:

<img width="1353" alt="Screenshot 2019-09-20 at 10 34 12" src="https://user-images.githubusercontent.com/314716/65318473-8df89780-db95-11e9-9c68-0824e51979d8.png">

And what happens if I load the tour in the desktop version:

![windows](https://user-images.githubusercontent.com/314716/65318650-ed56a780-db95-11e9-8388-943ca9eae473.jpg)

The way I implemented this was that in ``AddFilesToCabinet`` I check if a size column and normalization are being used, and if so, I make a temporary copy of the table in which I add a new column that contains the normalized sizes (not that this doesn't apply any e.g. power-law scaling and so on, since that can still be applied by older clients). The name of the column that is added is a fixed UUID - the reason for using a UUID is to ensure we don't conflict with user columns, and the reason for it being fixed is to avoid the situation where we keep adding columns to the table.

The approach is then that if this column is added, ``SizeColumn`` is made to point at it, while ``NormalizeColumn`` points to the original column. Then clients that understand ``NormalizeColumn`` use that as the ``SizeColumn`` when loading from XML, and discard the pre-computed sizes. In order to make the above possible, I've extended the Table class to be able to add/remove columns as well as make copies of tables.

This PR is currently rebased on https://github.com/WorldWideTelescope/wwt-web-client/pull/250 which adds a minimal JS test suite, and this PR adds some comprehensive tests of the new Table and SpreadSheetLayer functionality. Having the tests really helped polish things up and meant that things worked first time out of the box when I experimented with loading things in to the desktop client! Once https://github.com/WorldWideTelescope/wwt-web-client/pull/250 is ready to go and merged, I can rebase this PR.

If you are happy with this approach, I can take a stab at implementing something similar for #238!

pkgw commented:
@astrofrog A side comment: I'm unable to view the test results since `peter.williams@aas.org` is not authorized to view them. Any way to make them world-readable?
